File: viewer/client_stream_microphone_array.py
#------------------------------------------------------------------------------
# This script receives RAW microphone audio from the HoloLens microphone array
# and plays it.
# Audio stream configuration is fixed to 5 channels, 48000 Hz, float32.
# All channels are received but only one channel is played at a time.
# Press esc to stop.
#------------------------------------------------------------------------------

from pynput import keyboard

import numpy as np
import hl2ss
import hl2ss_lnm
import hl2ss_utilities

# Settings --------------------------------------------------------------------

# HoloLens address
host = "192.168.1.7"

# Channel
# Options:
# hl2ss.Parameters_MICROPHONE.ARRAY_TOP_LEFT
# hl2ss.Parameters_MICROPHONE.ARRAY_TOP_CENTER
# hl2ss.Parameters_MICROPHONE.ARRAY_TOP_RIGHT
# hl2ss.Parameters_MICROPHONE.ARRAY_BOTTOM_LEFT
# hl2ss.Parameters_MICROPHONE.ARRAY_BOTTOM_RIGHT
channel = hl2ss.Parameters_MICROPHONE.ARRAY_TOP_LEFT

#------------------------------------------------------------------------------

if __name__ == '__main__':
    audio_subtype     = np.float32
    audio_planar      = False
    audio_channels    = 1 # audio_player only supports 1 or 2 channels
    audio_sample_rate = hl2ss.Parameters_MICROPHONE.SAMPLE_RATE

    listener = hl2ss_utilities.key_listener(keyboard.Key.esc)
    listener.open()

    player = hl2ss_utilities.audio_player(audio_subtype, audio_planar, audio_channels, audio_sample_rate)
    player.open()

    client = hl2ss_lnm.rx_microphone(host, hl2ss.StreamPort.MICROPHONE, profile=hl2ss.AudioProfile.RAW, level=hl2ss.AACLevel.L5)
    client.open()

    while (not listener.pressed()): 
        data = client.get_next_packet()
        # Extract channel to play
        data.payload = data.payload[:, channel::hl2ss.Parameters_MICROPHONE.ARRAY_CHANNELS]
        player.put(data.timestamp, data.payload)

    client.close()
    player.close()
    listener.close()

File: extensions/client_matlab/viewer/client_stream_microphone_array.m
%%
% This script records microphone audio from the HoloLens and plays it.

%% Settings

% HoloLens address
host = '192.168.1.7';

% Recording length in seconds
length = 10;

% Channel
% Options:
% hl2ss.parameters_microphone.ARRAY_TOP_LEFT
% hl2ss.parameters_microphone.ARRAY_TOP_CENTER
% hl2ss.parameters_microphone.ARRAY_TOP_RIGHT
% hl2ss.parameters_microphone.ARRAY_BOTTOM_LEFT
% hl2ss.parameters_microphone.ARRAY_BOTTOM_RIGHT
channel = hl2ss.parameters_microphone.ARRAY_TOP_LEFT;

%%

client = hl2ss.mt.sink_microphone(host, hl2ss.stream_port.MICROPHONE, 4096, hl2ss.audio_profile.RAW, hl2ss.aac_level.L5);
client.open();

packet_duration = hl2ss.parameters_microphone.GROUP_SIZE_RAW / hl2ss.parameters_microphone.SAMPLE_RATE;
packet_count = length / packet_duration;
record = [];
frame_index = -1; % -1 for most recent frame

try
while (packet_count > 0)
    data = client.get_packet_by_index(frame_index);
    if (data.status == 0) % got packet
        frame_index = data.frame_index + 1;
        record = [record, data.audio];
        packet_count = packet_count - 1;
    else % no data
        pause(packet_duration * 10); % wait for data
    end
end
catch ME
    disp(ME.message);
end

client.close();

% play recorded sound
sound(record(channel, :), hl2ss.parameters_microphone.SAMPLE_RATE);
